### PR TITLE
do not load publish workflow request listener if routing is not availble

### LINKED ISSUE #232: publish_workflow  enabled by default means requiring routing-bundle
On my tests to install each bundle as separate on a symfony/skeleton app i realized that this bundle misses one dependency. The `symfony/routing-bundle`. `publisch_workflow` enabled means `request_listener` is set, which requires `DynamicRouter` to exist. That chain makes the routing-bundle to be a hard depency. You can see it when doing a `compoer require symfony-cmf/core-bundle` on an empty symfony app.

### PR DESCRIPTION
**Pull request by dbu**
fix #232, fix #233

the publish workflow request listener depends on the dynamic router. (not only with the constants, but also on the behaviour of putting the route document and content document into the request attributes).

i think this change is no BC break. if you have routing available, the default or explicit `true` behaviour stays the same. if you do not have routing, the previous version refused to build the container with default or explicit `true`. with explicit `false` there is again no behaviour change.

### CODE REVIEW

**Comment by ElectricMaxxx** [comment]
When i run `./bin/install.sh -psymfony-cmf/core-bundle -vdev-routing-bundle-optional` in the [testing-script](https://github.com/symfony-cmf/install-testing) i got now:
```
!!    [Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException]   
!!    The service "cmf_core.publish_workflow.checker.default" has a dependency on  
!!     a non-existent service "security.token_storage". 
```

possibly not related to the routing-bundle problem, but the next dependency issue.

**Comment by ElectricMaxxx** [comment]
@dbu can you please add an `composer require "symfony/security-core:^2.8|^3.0"` That component was also removed from `framework-bundle` from version 3.2 on.

**Comment by ElectricMaxxx** [comment]
`symfony/security` could possibly enough.

**Comment by dbu** [comment]
i added the check for security existing to the criteria for the `auto` mode.

**Comment by ElectricMaxxx** [comment]
Will check your branch later, thanks.

File: src/DependencyInjection/CmfCoreExtension.php [hunk 1]
```diff
@@ -318,16 +319,31 @@ private function loadPublishWorkflow($config, XmlFileLoader $loader, ContainerBu
         $container->setParameter($this->getAlias().'.publish_workflow.view_non_published_role', $config['view_non_published_role']);
         $loader->load('publish-workflow.xml');
```
Comment (ElectricMaxxx) [inline src/DependencyInjection/CmfCoreExtension.php]: shouldn't this line go somewhere behind the checks? Or is the removing stuff enough?

I still got
```
!!    The service "cmf_core.publish_workflow.checker.default" has a dependency on  
!!  
!!  
!!     a non-existent service "security.token_storage".   
````
when installing your branch.

```
maximilian@maximilian-ThinkPad-T540p:~/OpenSource/Cmf/install-testing/build/symfony-cmf/core-bundle/dev-routing-bundle-optional/test-app$ composer show symfony-cmf/*
symfony-cmf/core-bundle dev-routing-bundle-optional 75fbb35 Symfony CMF Core Bundle
``

**Comment by dbu** [comment]
ah, crap. yeah, i guess all of the workflow thing requires security. i will move that part up to not load any workflow when symfony/security is missing.

**Comment by ElectricMaxxx** [comment]
What about requiring `symfony/security`?

**Comment by dbu** [comment]
core bundle is other things than publish workflow too, so i don't want to force it. also, the security bundle needs to actually be loaded in the kernel for things to work

**Comment by dbu** [comment]
just pushed an (untested) change that hopefully fixes the situation

**Comment by dbu** [comment]
no idea why this now fails. it installs symfony/symfony so the interface should exist

**Comment by ElectricMaxxx** [comment]
Looking at https://packagist.org/packages/symfony/symfony#v2.8.0 i see no security except acl.

File: src/DependencyInjection/CmfCoreExtension.php [hunk 2]
```diff
@@ -315,19 +312,37 @@ public function setupFormTypes(ContainerBuilder $container, LoaderInterface $loa
...
     {
+        if (false === $config['enabled']
+            || ('auto' === $config['enabled'] && !class_exists(TokenStorageInterface::class))
+        ) {
+            $loader->load('no-publish-workflow.xml');
+
+            return;
+        }
+
+        if (!class_exists(TokenStorageInterface::class)) {
```
Comment (ElectricMaxxx) [inline src/DependencyInjection/CmfCoreExtension.php]: You should use `interface_exists` when asking for an interface

Comment (ElectricMaxxx) [reply src/DependencyInjection/CmfCoreExtension.php]: but i think, this will end in an warning, when the interface isn't there. but we will see.

Comment (dbu) [reply src/DependencyInjection/CmfCoreExtension.php]: d'oh, thats why.

i just pushed with a change to check if SecurityBundle is in the list of kernel.bundles. that is what we really care about i guess.

File: src/DependencyInjection/CmfCoreExtension.php [hunk 3]
```diff
@@ -315,19 +312,39 @@ public function setupFormTypes(ContainerBuilder $container, LoaderInterface $loa
      */
     private function loadPublishWorkflow($config, XmlFileLoader $loader, ContainerBuilder $container)
     {
+        $bundles = $container->getParameter('kernel.bundles');
+
+        if (false === $config['enabled']
+            || ('auto' === $config['enabled'] && !array_key_exists('SecurityBundle', $bundles))
```
Comment (ElectricMaxxx) [inline src/DependencyInjection/CmfCoreExtension.php]: please add an extension test for that configuration switch.

File: src/DependencyInjection/CmfCoreExtension.php [hunk 4]
```diff
@@ -315,19 +312,39 @@ public function setupFormTypes(ContainerBuilder $container, LoaderInterface $loa
...
+        ) {
+            $loader->load('no-publish-workflow.xml');
+
+            return;
+        }
+
+        if (!array_key_exists('SecurityBundle', $bundles)) {
```
Comment (ElectricMaxxx) [inline src/DependencyInjection/CmfCoreExtension.php]: please add an extension test.

File: src/DependencyInjection/CmfCoreExtension.php [hunk 5]
```diff
@@ -20,6 +20,7 @@
 use Symfony\Component\DependencyInjection\Loader\XmlFileLoader;
 use Symfony\Component\DependencyInjection\Reference;
 use Symfony\Component\HttpKernel\DependencyInjection\Extension;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
```
Comment (ElectricMaxxx) [inline src/DependencyInjection/CmfCoreExtension.php]: this one can be removed then.

**Comment by dbu** [comment]
we seem to lack quite some tests on this bundle... i added the ones for the publish workflow loading so at least this new functionality is covered.

**Comment by ElectricMaxxx** [comment]
Cool. It is awesome. Thanks a lot for this fix.

**Comment by ElectricMaxxx** [comment]
I restarted https://travis-ci.org/symfony-cmf/install-testing and i hope that the `core-bundle` related issues are gone now.

**Pull request merged**